rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,1122 @@
+* Login rules. We will recommend replacing login rules with access lists that provide similar functionality.
+
+## User stories
+
+Let’s get back to the issues we outlined  in the start of this RFD and review how the new system will  help to resolve them.
+
+### Gradual scoping
+
+Most Teleport customers will start with all resources and roles in the cluster scope. 
+
+We will let them introduce resource groups gradually. Let’s create two resource groups, `prod` and `dev` with resource groups `west` and `lab`.
+
+Here is the resource groups hierarchy, where we will assume that mars and luna servers matched the assignments:
+
+```mermaid
+graph TD;
+    luna(Server luna)-->west;
+    west-->prod;
+    prod-->cluster;
+    lab-->dev;
+    mars(Server mars)-->lab;
+    dev-->cluster;
+```
+
+```yaml
+kind: resource_group
+metadata:
+  name: prod
+---
+kind: resource_group
+metadata:
+  name: west
+  parent_resource_group: prod
+  # would be nice if we could match on AWS specific right away with match_aws
+  match_aws:
+    account_id: aws-account-id
+    region: west
+---
+kind: resource_group
+metadata:
+  name: dev
+---
+kind: resource_group
+metadata:
+  name: lab
+spec:
+  parent_resource_group: dev
+  # here we will just match on labels
+  match_labels:
+    kinds: [node]
+    env: lab
+```
+
+This will let administrators create a resource hierarchy by  mapping computing resources using AWS metadata or labels.
+
+We will use this setup in the following examples.
+
+### SSH access to specific hosts
+
+The most prominent use-case is our over-engineered access role. We can keep this role as is. Today, it grants blanket access to any computing resource of Teleport.
+
+Alice, who is an administrator, would like to restrict access for a user bob@example.com to any server in the lab as root
+
+```yaml
+kind: access_list
+metadata:
+  name: access-to-lab
+spec:
+  grants: 
+    roles: [access]
+    traits:
+      'internal.logins' : 'root'
+  scopes: ['/dev/lab']
+  members:
+    - bob@example.com
+```
+
+Teleport will grant role access  and traits internal.logins: root to `bob@example.com`, but only when Bob would access servers in the resource group `/dev/lab`. 
+
+This grant will not be valid out of the scope of `/dev/lab`, so Bob won’t be able to SSH as root to any other servers.
+
+K8s access to specific clusters
+
+Teleport can autodiscover clusters echo and bravo with namespaces default and prod, creating the following resource group hierarchy:
+
+```
+/k8s/namespaces/prod/bravo
+/k8s/namespaces/prod/echo
+
+/k8s/namespaces/default/bravo
+/k8s/namespaces/default/echo
+```
+
+Note that here we have set namespaces, and not cluster names as the root of the resource hierarchy,
+so we can group different cluster names by namespace.
+
+We can then use this hierarchy to create access lists specifying access to default namespace in any cluster:
```

Comment:
@klizhentas, should we maintain the current synchronization behavior, or consider transitioning to a method similar to what I outlined in https://github.com/gravitational/teleport/pull/38078#discussion_r1489463044?

The problem highlighted in https://github.com/gravitational/teleport/pull/38078#discussion_r1489463044 is significant because Teleport extends support for precise permissions beyond Kubernetes namespaces to include items such as secrets, pods, and configmaps.

The use of fine-grained permissions typically involves either providing access exclusively to specific pods within the same namespace or restricting access to secrets and configmaps within that namespace. The existing synchronization behavior poses challenges when applying fine-grained permissions to pods, as it requires syncing all objects, and pod names may be random, causing compatibility issues with our backends.

I understand why we consider this to represent fine grained stuff in TAG but I am afraid of effects in our backends

Reply:
I think we actually should use trait grants and existing templates in RBAC to avoid creating new concept.  I will clarify.